File: merge-two-sorted-lists/v2.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    var head: ListNode? = null
    var node: ListNode? = null

    fun setNextNode(next: ListNode?) {
        if (head == null) {
            head = next
        }
        node?.next = next
        node = next
    }

    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        var a = list1
        var b = list2

        while (a != null || b != null) {
            if (a != null && b != null) {
                if (a.`val` < b.`val`) {
                    setNextNode(a)
                    a = a.next
                } else {
                    setNextNode(b)
                    b = b.next
                }
            } else if (a != null) {
                setNextNode(a)
                a = a.next
            } else if (b != null) {
                setNextNode(b)
                b = b.next
            }
        }

        node?.next = null
        return head
    }
}

// Runtime: 290 ms, faster than 34.25% of Kotlin online submissions for Merge Two Sorted Lists.
// Memory Usage: 36.2 MB, less than 29.80% of Kotlin online submissions for Merge Two Sorted Lists.

File: merge-sorted-array/v2.go
package v2

func merge(nums1 []int, m int, nums2 []int, n int) {
	idx1 := m - 1
	idx2 := n - 1
	nm := n + m
	for i := 0; i < nm; i++ {
		if idx1 < 0 || (idx1 >= 0 && idx2 >= 0 && nums1[idx1] < nums2[idx2]) {
			nums1[nm-1-i] = nums2[idx2]
			idx2--
		} else {
			nums1[nm-1-i] = nums1[idx1]
			idx1--
		}
	}
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Merge Sorted Array.
// Memory Usage: 2.3 MB, less than 21.74% of Go online submissions for Merge Sorted Array.
